core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java

```diff
@@ -86,67 +87,81 @@ public HttpResponse execute(ClientRequestContext ctx, HttpRequest req) throws Ex
             return HttpResponse.ofFailure(cause);
         }
 
-        final Endpoint endpointWithPort = endpoint.withDefaultPort(ctx.sessionProtocol().defaultPort());
+        final Endpoint endpointWithPort;
+        final int defaultPort = ctx.sessionProtocol().defaultPort();
+        final String host = extractHostFromAuthority(ctx, req);
+        if (host == null) {
+            endpointWithPort = endpoint.withDefaultPort(defaultPort);
+        } else {
+            // Override the host in endpoint with the extracted host.
+            if (endpoint.hasPort()) {
+                endpointWithPort = Endpoint.of(host, endpoint.port());
```

Comment:
Isn't it possible to tweak this to create endpoints without validation somehow? Better to do that then add a cache I think - cache has its own overhead so should only use when needed.

Reply:
> Isn't it possible to tweak this to create endpoints without validation somehow?

Added an additional condition to compare the extracted host with endpoint's host.
This branch will be entered only when users set an additional `AUTHORITY` header.
https://github.com/line/armeria/blob/baf299c00a59392620fabf133da6bee066331033/core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java#L228-L229

However, I think a cache is still valuable for a `WebClient` created without a based URI. `Endpoint.parse()` will be invoked for every request.
https://github.com/line/armeria/blob/369b7f2fb9631db77d730891a49d2077b6c31c27/core/src/main/java/com/linecorp/armeria/client/DefaultWebClient.java#L76

Reply:
Hmm - perhaps the cache should be kept for just that codepath? I feel there are a lot of unknown consequences of caching Endpoint generally like this. In particular, it seems like a dynamic endpoint group will fill up the cache with a bunch of IP addresses.

Reply:
That makes sense. 1) Removed the cache for `Endpoint.of()` and 2) added `Endpoint.unsafeCreate()` that applied it to where host is already valid.